Fix build_ignore when installing a collection from source

s-hertel opened the pull request:
##### SUMMARY
Only create the directory in question (as opposed to copying the tree), since the contents of the source dir may be in the the collection's `build_ignore` metadata. Also sort the FILES.json by name so we're sure to create directories before any files contained.

Fixes #75528

##### ISSUE TYPE
- Bugfix Pull Request

s-hertel committed:
Make individual dirs that should exist instead of using shutil.copytree

s-hertel committed:
Test build_ignore when installing collections in git repos

s-hertel committed:
changelog

s-hertel committed:
Fix assertion
Fix git directory

webknjaz reviewed (commented):
Could you add a test for installing the collection from source vs build artifact in different directories and running `diff -ur` on them to verify that the contents are exactly the same?

s-hertel commented:
@webknjaz We don't have any tests for local folders yet, so I'm going to do it as a follow-up. It was easy just to modify the SCM tests quickly since it's a common code path.

webknjaz commented:
> Oh - maybe you didn't specifically mean from a source dir. I can add another test that builds the artifact and installs with both methods, good idea :+1:

Yep, this is exactly what I was thinking. Sorry, I wasn't clear about this.

ssbarnea commented:
I still see deep `__pycache__` folders and `.DS_Store` files being copied to the destination but not being listed inside `MANIFEST.json`.

I tried to exclude them adding extra `*/__pycache__` entry to the `build_ignore` list but to no use.

s-hertel committed:
Fix nested content while building the collection dir

Test installing a collection from a git repo and artifact have the same result

Refactor to use variables

s-hertel commented:
@ssbarnea Thanks very much for taking a look.

> I still see deep `__pycache__` folders and .DS_Store files being copied to the destination but not being listed inside MANIFEST.json.

I'm not able to reproduce on this branch. Do you have a minimal example so I can add another test case for it?

s-hertel commented:
/rebuild_merge

mkrizek commented:
rebuild_merge

ssbarnea commented:
@s-hertel I wonder how hard it would be to backport this fix to 2.11, so people building collections can benefit from it on next bugfix release.

s-hertel commented:
@ssbarnea I opened a backport PR: #75623
